presto-main/src/main/java/com/facebook/presto/memory/HighMemoryTaskKiller.java

```diff
@@ -0,0 +1,248 @@
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private boolean shouldTriggerTaskKiller(GarbageCollectionNotificationInfo info)
+    {
+        boolean triggerTaskKiller = false;
+        DataSize beforeGcDataSize = info.getBeforeGcTotal();
+        DataSize afterGcDataSize = info.getAfterGcTotal();
+
+        if (taskKillerStrategy == FREE_MEMORY_ON_FREQUENT_FULL_GC) {
+            long currentGarbageCollectedBytes = beforeGcDataSize.toBytes() - afterGcDataSize.toBytes();
+            long currentFullGCTimestamp = ticker.read();
+
+            if (isFrequentFullGC(lastFullGCTimestamp, currentFullGCTimestamp) && !hasFullGCFreedEnoughBytes(currentGarbageCollectedBytes)) {
+                triggerTaskKiller = true;
+            }
+
+            lastFullGCTimestamp = currentFullGCTimestamp;
+            lastFullGCCollectedBytes = currentGarbageCollectedBytes;
+        }
+        else if (taskKillerStrategy == FREE_MEMORY_ON_FULL_GC) {
+            if (isLowMemory() && beforeGcDataSize.toBytes() - afterGcDataSize.toBytes() < reclaimMemoryThreshold) {
+                triggerTaskKiller = true;
+            }
+        }
+
+        return triggerTaskKiller;
+    }
+
+    private List<SqlTask> getActiveTasks()
+    {
+        return sqlTaskManager.getAllTasks().stream()
+                .filter(task -> !task.getTaskState().isDone())
+                .collect(toImmutableList());
+    }
+
+    @VisibleForTesting
+    public static Optional<QueryId> getMaxMemoryConsumingQuery(ListMultimap<QueryId, SqlTask> queryIDToSqlTaskMap)
+    {
+        if (queryIDToSqlTaskMap.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Comparator<Map.Entry<QueryId, Long>> comparator = Comparator.comparingLong(Map.Entry::getValue);
+
+        Optional<QueryId> maxMemoryConsumpingQueryId = queryIDToSqlTaskMap.asMap().entrySet().stream()
+                //Convert to Entry<QueryId, Long>, QueryId -> Total Memory Reservation
+                .map(entry ->
+                        new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue().stream()
+                                .map(SqlTask::getTaskInfo)
+                                .map(TaskInfo::getStats)
+                                .mapToLong(stats -> stats.getUserMemoryReservationInBytes() + stats.getSystemMemoryReservationInBytes() + stats.getRevocableMemoryReservationInBytes())
+                                .sum())
+                ).max(comparator.reversed()).map(Map.Entry::getKey);
+
+        return maxMemoryConsumpingQueryId;
+    }
+
+    private boolean isFrequentFullGC(long lastFullGCTimestamp, long currentFullGCTimestamp)
+    {
+        long diffBetweenFullGCMilis = (currentFullGCTimestamp - lastFullGCTimestamp) / 1000000;
```

Comment:
Idk, https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/TimeUnit.html#convert-long-java.util.concurrent.TimeUnit- ? Or the airlift Duration ?

Reply:
Done.